src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -259,417 +258,481 @@ AND NOT EXISTS (
+                predicates.add(builder.greaterThanOrEqualTo(root.get(Pool_.endDate), activeOn));
             }
             else {
-                criteria.add(Restrictions.ge("Pool.endDate", activeOn));
+                predicates.add(builder.greaterThanOrEqualTo(root.get(Pool_.endDate), activeOn));
             }
         }
 
         if (after != null) {
-            criteria.add(Restrictions.gt("Pool.startDate", after));
+            predicates.add(builder.greaterThanOrEqualTo(root.get(Pool_.startDate), after));
         }
 
-        // TODO: This section is sloppy. If we're going to clobber the bits in the filter with our own input
-        // parameters, why bother accepting a filter to begin with? Similarly, why bother accepting a filter
-        // if the method takes the arguments directly? If we're going to abstract out the filtering bits, we
-        // should go all-in, cut down on the massive argument list and simply take a single filter object.
-        // -C
+        if (predicates.isEmpty()) {
+            return null;
+        }
+
+        return builder.and(predicates.toArray(new Predicate[0]));
+    }
+
+    private Predicate getSubscriptionPredicate(Root<Pool> root, CriteriaBuilder builder,
+        Set<String> subscriptionIds) {
+
+        if (subscriptionIds == null || subscriptionIds.isEmpty()) {
+            return null;
+        }
+
+        Join<Pool, SourceSubscription> subJoin = root.join(Pool_.sourceSubscription);
+
+        return inPredicate(builder, subJoin.get(SourceSubscription_.subscriptionId), subscriptionIds);
+    }
+
+    private Predicate getProductsPredicate(CriteriaBuilder builder, Join<Pool, Product> product,
+        SetJoin<Product, Product> providedProduct, Set<String> productIds, Set<String> productNames) {
+
+        if ((productIds == null || productIds.isEmpty()) &&
+            (productNames == null || productNames.isEmpty())) {
+            return null;
+        }
 
-        if (subscriptionId != null && !subscriptionId.isEmpty()) {
-            criteria.createAlias("Pool.sourceSubscription", "srcsub")
-                .add(Restrictions.eq("srcsub.subscriptionId", subscriptionId));
+        if (providedProduct == null) {
+            providedProduct = product.join(Product_.providedProducts, JoinType.LEFT);
         }
 
-        boolean joinedProvided = false;
+        List<Predicate> predicates = new ArrayList<>();
         if (productIds != null && !productIds.isEmpty()) {
-            if (!joinedProvided) {
-                criteria.createAlias("Product.providedProducts", "Provided", JoinType.LEFT_OUTER_JOIN);
-                joinedProvided = true;
-            }
+            predicates.add(inPredicate(builder, product.get(Product_.id), productIds));
+            predicates.add(inPredicate(builder, providedProduct.get(Product_.id), productIds));
+        }
+
+        if (productNames != null && !productNames.isEmpty()) {
```

Comment:
I'm not seeing product names in the original query. Is this a net new addition or something for other operations elsewhere in this curator?

Reply:
This is similar to the contractNumbers comment, this should be handled by the `getMatchesPredicates` method

Reply:
We can also get rid of `orderNumbers` from the `PoolQualifier` as well!